### DISCUSSION TITLE
Help wanted - How to apply my own filter to sprite that is used as a mask?

### DISCUSSION THREAD

**Discussion by Kalsyc** [pixijs/pixijs#7520] [Q&A]
### Expected Behavior

Here's what I'm trying to do: I have a background image in which it is blurred, and when the user moves around the screen (either via mouse or touch), the picture will appear clear. I can simply do this by applying a blur filter to the outside background image whilst applying a circle as a mask on another image that is clear.

However trouble comes within the circle as there are areas in which the masked out areas will appear blurry or distorted (depending on what filter I place in). Note that the blur filter I wish to place inside these masked out areas within the circle is different from the outside blur and it must be contained within the quadrant-shaped masks and not the entire image.

Here is a visualisation of what i want to do:
![image](https://user-images.githubusercontent.com/45655404/120276542-69027480-c2e5-11eb-89df-44d971cd6d4c.png)

### Current Behavior

Currently, applying the filter to the quadrant-shaped masked areas (in which is a Sprite object) doesn't do anything. the only way the blur works is if I apply it to the background sprite or the container, however this isn't what I want as I want to localize the filter on the masked area.

In other words, the blur filter (and whatever filter I add in) is only applied to the entire image rather than applying it only on the masked area.

### Possible Solution

I have tried using [this](https://pixijs.io/examples/#/masks/filter.js) but to no avail as it seems like the filter only acts on the blending between the sprite and the bounds and not the entire masked sprite.

Anyone can help me with this? I'm not so sure if I'm even approaching this the right way. Thank you in advance!

### Environment

- **`pixi.js` version**: _e.g. 6 (latest)_
- **Browser & Version**: _e.g. Safari v14.1_
- **OS & Version**: _e.g. MacOS Big Sur (M1)_
- **Running Example**: _url_

**Comment by ivanpopelyshev** [comment]
In my apps I use custom filter and I know the hack how to make it right :) Basically, instead of multiple passes with same strength, you have to do passes with strength `s, s*2, s*4, s*8` . For example with kernel 5 it allows blur strength up to 256 after 16 passes (8 horiz, 8 vert). 

What strength/quality/kernel do you use currently? Which params for same CSS blur you want?
I can publish that filter after a bit of code polishing. How serious do you need it?

**Comment by Kalsyc** [comment]
Thanks for the quick response!

Actually, the outside blur isn't as important, what is important is that I need to be able to apply a filter to a masked area of the background image. This is the current blur I'm experimenting with which is a single pass of Gaussian Blur in some x/y direction but with a linear gradient from `min_sigma` to `max_sigma`. The kernel isn't that important as long as I'm able to change it within the js script.

```
// Horizontal (X) Blur Fragment Shader (Left to Right)

uniform float pixelSize; // size of pixel (width)
uniform float kernel; // radius size of kernel
uniform float min_sigma; // minimum intensity of blur
uniform float max_sigma; // maximum intensity of blur

varying vec2 vTextureCoord;
uniform vec4 dimensions;
uniform sampler2D uSampler;

const float pi = 3.14159265;
const float maxPixelsPerSide = 15.0; // Fallback

// Function that calculates the weight of the pixel
float calculateWeight(float offset, float sd) {
  if (sd <= 0.000001) {
    sd = 0.000001;
  }
  float circle_denom = 1.0 / (sqrt(2.0 * pi * sd * sd));
  float expo = exp(-((offset * offset) / (2.0 * sd * sd)));
  float result = circle_denom * expo;
  if (result > 1.0) {
    return 1.0;
  } else {
    return result;
  }
}

void main(void) {
  vec4 avgValue = vec4(0.0, 0.0, 0.0, 0.0);
  float sd = (max_sigma - min_sigma) * vTextureCoord.x;
  float remainder = 1.0;
  float current;
  for (float i = 1.0; i <= maxPixelsPerSide; i++) {
    if (i > kernel) {
      break;
    }
    current = calculateWeight(i, sd);
    avgValue += texture2D(uSampler, vec2(vTextureCoord.x - i * pixelSize, vTextureCoord.y)) * current;   
    avgValue += texture2D(uSampler, vec2(vTextureCoord.x + i * pixelSize, vTextureCoord.y)) * current;
    remainder -= current * 2.0;
  }
  avgValue +=  texture2D(uSampler, vTextureCoord.st) * remainder;
  gl_FragColor = avgValue;

}
```
```
// Default vert shader

attribute vec2 aVertexPosition;
attribute vec2 aTextureCoord;
uniform mat3 projectionMatrix;
varying vec2 vTextureCoord;

void main(void)
{
    gl_Position = vec4((projectionMatrix * vec3(aVertexPosition, 1.0)).xy, 0.0, 1.0);
    vTextureCoord = aTextureCoord;
}
```

Thus, it'll look like this if I use a rectangle image (which will be hard to do if I want to apply it on a curve I guess, but right here I'm applying the horizontal blur filter pass in both x and y directions with a linear gradient) Just imagine if this is localised onto the masked quadrant-shaped area of a lens in which the user can move around the display:
![image](https://user-images.githubusercontent.com/45655404/120427531-a417ad00-c3a4-11eb-912b-f760257990c5.png)

Anyways, are you also saying that if I were to do something like that, I would need to write a custom filter for the entire image? How would that work if I had to input mouse coordinates? Thanks for your help in advance :)

**Comment by ivanpopelyshev** [comment]
`import('@pixi/picture').MaskFilter` works for any base filter: https://pixijs.io/examples/#/plugin-picture/mask-blur.js

just re-published the plugin for v6 yesterday.
